pkg/controller/pxc/controller_test.go

```diff
@@ -540,3 +541,749 @@ var _ = Describe("Finalizer delete-pxc-pvc", Ordered, func() {
+			Name:      ns,
+			Namespace: ns,
+		},
+	}
+
+	BeforeAll(func() {
+		By("Creating the Namespace to perform the tests")
+		err := k8sClient.Create(ctx, namespace)
+		Expect(err).To(Not(HaveOccurred()))
+	})
+
+	AfterAll(func() {
+		By("Deleting the Namespace to perform the tests")
+		_ = k8sClient.Delete(ctx, namespace)
+	})
+
+	Context("Cluster is deployed with ProxySQL", Ordered, func() {
+		const crName = "auth-policy-proxysql"
+		crNamespacedName := types.NamespacedName{Name: crName, Namespace: ns}
+
+		cr, err := readDefaultCR(crName, ns)
+		It("should read default cr.yaml", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should create PerconaXtraDBCluster", func() {
+			cr.Spec.HAProxy.Enabled = false
+			cr.Spec.ProxySQL.Enabled = true
+
+			Expect(k8sClient.Create(ctx, cr)).Should(Succeed())
+		})
+
+		It("should reconcile", func() {
+			_, err := reconciler().Reconcile(ctx, ctrl.Request{NamespacedName: crNamespacedName})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should use mysql_native_password", func() {
+			sts := appsv1.StatefulSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      crName + "-pxc",
+					Namespace: ns,
+				},
+			}
+			err := k8sClient.Get(ctx, client.ObjectKeyFromObject(&sts), &sts)
+			Expect(err).NotTo(HaveOccurred())
+
+			envFound := false
+			for _, c := range sts.Spec.Template.Spec.Containers {
+				if c.Name == "pxc" {
+					for _, e := range c.Env {
+						if e.Name == "DEFAULT_AUTHENTICATION_PLUGIN" {
+							envFound = true
+							Expect(e.Value).To(Equal("mysql_native_password"))
+						}
+					}
+				}
+			}
+
+			Expect(envFound).To(BeTrue())
+		})
+	})
+
+	Context("Cluster is deployed with HAProxy", Ordered, func() {
+		const crName = "auth-policy-haproxy"
+		crNamespacedName := types.NamespacedName{Name: crName, Namespace: ns}
+
+		cr, err := readDefaultCR(crName, ns)
+		It("should read default cr.yaml", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should create PerconaXtraDBCluster", func() {
+			cr.Spec.HAProxy.Enabled = true
+			cr.Spec.ProxySQL.Enabled = false
+
+			Expect(k8sClient.Create(ctx, cr)).Should(Succeed())
+		})
+
+		It("should reconcile", func() {
+			_, err := reconciler().Reconcile(ctx, ctrl.Request{NamespacedName: crNamespacedName})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should use caching_sha2_password", func() {
+			sts := appsv1.StatefulSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      crName + "-pxc",
+					Namespace: ns,
+				},
+			}
+			err := k8sClient.Get(ctx, client.ObjectKeyFromObject(&sts), &sts)
+			Expect(err).NotTo(HaveOccurred())
+
+			envFound := false
+			for _, c := range sts.Spec.Template.Spec.Containers {
+				if c.Name == "pxc" {
+					for _, e := range c.Env {
+						if e.Name == "DEFAULT_AUTHENTICATION_PLUGIN" {
+							envFound = true
+							Expect(e.Value).To(Equal("caching_sha2_password"))
+						}
+					}
+				}
+			}
+
+			Expect(envFound).To(BeTrue())
+		})
+
```

Comment:
These last two `It` blocks I would put in a when block like:
`When("Proxy is switched from HAProxy to ProxySQL", func (){...})`

Think it would be more clear.

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/1415/commits/fe788b6b8ffc67beebb2f311183a59fd72ab463f